File: cannabis_management/cannabis_management/page/ar_legacy/ar_legacy.py
"""AR Legacy — legacy receivables grouped by who owns the relationship.

The legacy book (invoices posted on or before the AR cut-over) is being worked
through account by account, and the first question on every one of them is whose
relationship it is. This page answers that: every legacy customer with money
still outstanding, filed under one of the segments below, with the balance and
invoice count so the size of each pile is obvious.

The segment is stored on the Customer itself (custom_ar_legacy_segment), so it
survives, reports and filters like any other field rather than living only in
this page.

Numbers come from the same source as the AR dashboard — Sales Invoice
outstanding for invoices up to LEGACY_CUTOFF, intercompany customers excluded —
so the two pages can never disagree about what legacy AR is.
"""

import frappe
from frappe.utils import date_diff, flt, getdate, nowdate

from cannabis_management.cannabis_management.page.ar_dashboard.ar_dashboard import (
	LEGACY_CUTOFF,
	NEW_AR_START,
	ORG_WIDE_VIEW_ROLES,
	_internal_customer_names,
	_permitted_companies,
)

# The original five. No longer the source of truth - that is the "AR Segment"
# doctype, one record per segment - but kept as the seed for a fresh site and as
# the fallback when the doctype is not installed, so the page cannot break
# between deploying code and migrating.
SEGMENTS = [
	"Matt's Accounts",
	"Matt has Relationships",
	"Nikki's Accounts - Matt knows them",
	"Nikki's Accounts - Legit companies",
	"Nikki's Accounts - Unknown companies",
]

AR_SEGMENT_DOCTYPE = "AR Segment"

# Which invoices each mode counts. The segment itself is a Customer attribute and
# is NOT per-mode: switching modes changes the money on the page, not who owns
# the relationship.
AR_MODES = ("legacy", "new")


def get_segments():
	"""Segments offered on the page, in display order.

	``[{"value": <segment>, "color": <hex or None>}]``. The blank "Unassigned"
	entry is added by the front end; it is not a record.
	"""
	fallback = [{"value": s, "color": None} for s in SEGMENTS]

	if not frappe.db.exists("DocType", AR_SEGMENT_DOCTYPE):
		return fallback

	rows = frappe.get_all(
		AR_SEGMENT_DOCTYPE,
		filters={"disabled": 0},
		fields=["name", "color"],
		order_by="display_order asc, creation asc",
	)
	if not rows:
		# No segments at all is never what anyone wants - treat it as
		# "not configured yet" and offer the originals.
		return fallback

	return [{"value": r["name"], "color": r.get("color") or None} for r in rows]


def sync_segment_field_options(exclude=None):
	"""Point the Customer form's own Select at the current segment list.

	Called from the AR Segment controller so the Customer field and this page can
	never disagree. ``exclude`` drops a record that is mid-delete and therefore
	still readable.
	"""
	values = [s["value"] for s in get_segments() if s["value"] != exclude]

	name = frappe.db.get_value(
		"Custom Field", {"dt": "Customer", "fieldname": SEGMENT_FIELD}, "name"
	)
	if name:
		frappe.db.set_value("Custom Field", name, "options", "\n" + "\n".join(values))
		frappe.clear_cache(doctype="Customer")


def seed_segments():
	"""Create a record per original segment. Idempotent; safe to re-run.

	Colours match what ar_legacy.css already hard-codes for these five, so the
	page looks exactly as it did before they became records.
	"""
	if not frappe.db.exists("DocType", AR_SEGMENT_DOCTYPE):
		return

	colors = {
		"Matt's Accounts": "#7c3aed",
		"Matt has Relationships": "#2563eb",
		"Nikki's Accounts - Matt knows them": "#0d9488",
		"Nikki's Accounts - Legit companies": "#16a34a",
		"Nikki's Accounts - Unknown companies": "#d97706",
	}

	for i, segment in enumerate(SEGMENTS):
		if frappe.db.exists(AR_SEGMENT_DOCTYPE, segment):
			continue
		frappe.get_doc({
			"doctype": AR_SEGMENT_DOCTYPE,
			"segment_name": segment,
			"color": colors.get(segment),
			"display_order": (i + 1) * 10,
		}).insert(ignore_permissions=True)

UNASSIGNED = "Unassigned"

SEGMENT_FIELD = "custom_ar_legacy_segment"

# Who may move an account between segments. Viewing follows the page's own role
# gate; filing an account under someone's name is an ownership call, so it is
# kept to the people who own the AR process.
EDIT_ROLES = ORG_WIDE_VIEW_ROLES + ("Account Manager",)


def _can_edit(user=None):
	user = user or frappe.session.user
	if user == "Administrator":
		return True
	return bool(set(frappe.get_roles(user)).intersection(EDIT_ROLES))


@frappe.whitelist()
def get_data(ar_mode="legacy"):
	"""Every customer still owing money in the chosen book, with segment and totals.

	Read with raw SQL rather than get_list: the figures are the same for everyone
	who can open the page, and the page is role-gated already. Company scoping is
	applied explicitly from the caller's own User Permissions so a rep restricted
	to two entities does not see the rest.
	"""
	permitted = _permitted_companies()
	internal = _internal_customer_names()

	# The segment column is created by install_segment_field, not by migrate. On a
	# site where that has not been run yet, selecting it would fail with "Unknown
	# column" and the page would show nothing at all — so fall back to reading no
	# segment and let the page say what to run.
	has_field = frappe.db.has_column("Customer", SEGMENT_FIELD)
	segment_select = f"MAX(c.{SEGMENT_FIELD})" if has_field else "NULL"

	# Legacy is everything up to the cut-over, New AR everything from the day
	# after it - the same two windows the AR dashboard uses, so the pages agree.
	# Anything unrecognised falls back to legacy: that was the only behaviour
	# before the toggle existed and it must stay the default.
	ar_mode = ar_mode if ar_mode in AR_MODES else "legacy"

	conditions = ["si.docstatus = 1", "si.outstanding_amount > 0"]
	if ar_mode == "new":
		conditions.append("si.posting_date >= %(start)s")
		params = {"start": NEW_AR_START}
	else:
		conditions.append("si.posting_date <= %(cutoff)s")
		params = {"cutoff": LEGACY_CUTOFF}

	if internal:
		conditions.append("si.customer NOT IN %(internal)s")
		params["internal"] = tuple(internal)

	if permitted is not None:
		conditions.append("si.company IN %(companies)s")
		params["companies"] = tuple(permitted) or ("__none__",)

	rows = frappe.db.sql(
		f"""
		SELECT
			si.customer                                    AS customer,
			MAX(c.customer_name)                           AS customer_name,
			COUNT(*)                                       AS invoices,
			SUM(si.outstanding_amount * IFNULL(si.conversion_rate, 1)) AS outstanding,
			MIN(si.posting_date)                           AS oldest,
			{segment_select}                               AS segment,
			GROUP_CONCAT(DISTINCT si.company ORDER BY si.company SEPARATOR ', ') AS companies
		FROM `tabSales Invoice` si
		LEFT JOIN `tabCustomer` c ON c.name = si.customer
		WHERE {" AND ".join(conditions)}
		GROUP BY si.customer
		ORDER BY outstanding DESC
		""",
		params,
		as_dict=True,
	)

	today = getdate(nowdate())
	companies = set()
	for row in rows:
		row["outstanding"] = flt(row["outstanding"])
		row["segment"] = row.get("segment") or ""
		row["companies"] = row.get("companies") or ""
		oldest = row.get("oldest")
		row["oldest"] = str(oldest or "")
		# Age of the oldest unpaid invoice, in days — the useful sort for a
		# legacy book, where "how long has this been sitting" beats size.
		row["age_days"] = date_diff(today, getdate(oldest)) if oldest else 0
		for company in (row["companies"] or "").split(", "):
			if company:
				companies.add(company)

	segments = get_segments()

	return {
		"rows": rows,
		"companies": sorted(companies),
		"segments": [s["value"] for s in segments],
		"segment_colors": {s["value"]: s["color"] for s in segments if s["color"]},
		"unassigned_label": UNASSIGNED,
		"can_edit": _can_edit() and has_field,
		"cutoff": LEGACY_CUTOFF,
		"ar_mode": ar_mode,
		"field_missing": not has_field,
	}


@frappe.whitelist()
def set_segment(customer, segment):
	"""File one customer under a segment. Blank moves it back to Unassigned."""
	if not _can_edit():
		frappe.throw(
			"You do not have permission to move accounts between segments.",
			frappe.PermissionError,
		)

	segment = (segment or "").strip()
	# Checked against the live list, so a segment added today is accepted today.
	if segment and segment not in [s["value"] for s in get_segments()]:
		frappe.throw(f"Unknown segment: {segment}")

	if not frappe.db.has_column("Customer", SEGMENT_FIELD):
		frappe.throw(
			"The AR Legacy segment field is not installed on this site yet. Run: "
			"bench --site &lt;site&gt; execute cannabis_management.cannabis_management."
			"page.ar_legacy.ar_legacy.install_segment_field"
		)

	if not frappe.db.exists("Customer", customer):
		frappe.throw(f"Customer {customer} not found")

	frappe.db.set_value("Customer", customer, SEGMENT_FIELD, segment)
	return {"customer": customer, "segment": segment}


def install_segment_field():
	"""Create the Customer field this page writes to. Idempotent — safe to re-run.

	Deliberately NOT added to the app's Custom Field fixture: that fixture is
	re-imported by every bench migrate and would fight any later change to this
	field. Adding a new field is safe from it — the fixture only overwrites what
	it already contains — so this installer is the only thing that owns it.
	"""
	from frappe.custom.doctype.custom_field.custom_field import create_custom_fields

	create_custom_fields(
		{
			"Customer": [
				{
					"fieldname": SEGMENT_FIELD,
					"fieldtype": "Select",
					"label": "AR Legacy Segment",
					"options": "\n" + "\n".join([s["value"] for s in get_segments()]),
					"insert_after": "custom_reconciliation_status",
					"in_standard_filter": 1,
					"description": "Set from the AR Legacy page — whose relationship this legacy account is.",
				}
			]
		},
		update=True,
	)
	frappe.db.commit()
	return {"status": "ok", "field": SEGMENT_FIELD, "segments": SEGMENTS}

File: cannabis_management/cannabis_management/page/ar_legacy/ar_legacy.js
// AR Legacy — legacy receivables filed by whose relationship the account is.
//
// Layout, in the order it is read:
//   1. KPI cards — the size of the book, and how much of it is still unfiled.
//   2. Segment cards — count and value per segment; click one to filter to it.
//   3. Filters — search, company, age, minimum balance, sort.
//   4. Sections — one per segment, each with its own subtotal.
//
// Changing a row's segment re-parents it immediately and recomputes every
// total on the page, so the effect of the choice is visible without a reload.

frappe.pages['ar-legacy'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'AR Legacy',
		single_column: true
	});

	wrapper.page = page;

	var state = {
		rows: [], segments: [], companies: [], can_edit: false,
		filters: { search: '', segment: '', company: '', age: '', min_amount: '', sort: 'outstanding' },
		collapsed: {}
	};
	page._state = state;

	var AGE_BUCKETS = [
		{ value: '', label: 'Any age' },
		{ value: '90', label: '90+ days' },
		{ value: '180', label: '180+ days' },
		{ value: '270', label: '270+ days' },
		{ value: '365', label: '1 year+' }
	];

	var SORTS = [
		{ value: 'outstanding', label: 'Largest balance' },
		{ value: 'age', label: 'Oldest first' },
		{ value: 'invoices', label: 'Most invoices' },
		{ value: 'name', label: 'Customer A–Z' }
	];

	page.main.html(`
		<div class="arl">
			<div class="arl-cards" id="arl-cards"></div>
			<div class="arl-seg-cards" id="arl-seg-cards"></div>

			<div class="arl-filters">
				<div class="arl-field arl-field-search">
					<span class="arl-search-icon">&#128269;</span>
					<input type="text" id="arl-search" class="arl-input" placeholder="Search customer…" />
				</div>
				<div class="arl-field">
					<label class="arl-label">Company</label>
					<select id="arl-company" class="arl-input"></select>
				</div>
				<div class="arl-field">
					<label class="arl-label">Age of oldest invoice</label>
					<select id="arl-age" class="arl-input"></select>
				</div>
				<div class="arl-field">
					<label class="arl-label">Min balance</label>
					<input type="number" id="arl-min" class="arl-input" placeholder="0" min="0" step="1000" />
				</div>
				<div class="arl-field">
					<label class="arl-label">Sort by</label>
					<select id="arl-sort" class="arl-input"></select>
				</div>
				<button class="arl-btn arl-btn-ghost" id="arl-clear">Clear</button>
				<button class="arl-btn" id="arl-refresh">&#8635; Refresh</button>
			</div>

			<div class="arl-active" id="arl-active"></div>

			<div id="arl-body">
				<div class="arl-loading"><div class="arl-spinner"></div><p>Loading legacy accounts…</p></div>
			</div>
		</div>
	`);

	// ---- helpers -------------------------------------------------------
	function money(v, compact) {
		v = v || 0;
		if (compact && Math.abs(v) >= 1000) {
			return '$' + (v / 1000).toFixed(v >= 100000 ? 0 : 1) + 'k';
		}
		return (v < 0 ? '-$' : '$') + Math.abs(v).toLocaleString('en-US',
			{ minimumFractionDigits: 2, maximumFractionDigits: 2 });
	}

	function esc(s) { return frappe.utils.escape_html(s == null ? '' : String(s)); }
	function slug(s) { return (s || 'unassigned').replace(/[^a-z0-9]+/gi, '-').toLowerCase(); }
	function label_of(seg) { return seg || 'Unassigned'; }
	function sum_of(rows) { return rows.reduce(function (a, r) { return a + (r.outstanding || 0); }, 0); }

	// Segment order: Unassigned first — an unfiled account is the thing that
	// still needs a decision — then the segments as declared server-side.
	function all_segments() { return [''].concat(state.segments); }

	// Everything except the segment filter, so the segment cards can show true
	// counts while one of them is selected.
	function rows_before_segment() {
		var f = state.filters;
		var q = (f.search || '').toLowerCase();
		var min = parseFloat(f.min_amount) || 0;
		var age = parseInt(f.age, 10) || 0;

		return state.rows.filter(function (r) {
			if (q && (r.customer || '').toLowerCase().indexOf(q) === -1 &&
				(r.customer_name || '').toLowerCase().indexOf(q) === -1) return false;
			if (f.company && (r.companies || '').indexOf(f.company) === -1) return false;
			if (min && (r.outstanding || 0) < min) return false;
			if (age && (r.age_days || 0) < age) return false;
			return true;
		});
	}

	function visible_rows() {
		var f = state.filters;
		var rows = rows_before_segment();
		if (f.segment === '__unassigned__') return rows.filter(function (r) { return !r.segment; });
		if (f.segment) return rows.filter(function (r) { return r.segment === f.segment; });
		return rows;
	}

	function sorted(rows) {
		var by = state.filters.sort;
		return rows.slice().sort(function (a, b) {
			if (by === 'age') return (b.age_days || 0) - (a.age_days || 0);
			if (by === 'invoices') return (b.invoices || 0) - (a.invoices || 0);
			if (by === 'name') {
				return (a.customer_name || a.customer).localeCompare(b.customer_name || b.customer);
			}
			return (b.outstanding || 0) - (a.outstanding || 0);
		});
	}

	// ---- cards ---------------------------------------------------------
	function render_cards() {
		var rows = visible_rows();
		var unassigned = rows.filter(function (r) { return !r.segment; });
		var assigned = rows.filter(function (r) { return !!r.segment; });
		var oldest = rows.reduce(function (a, r) { return Math.max(a, r.age_days || 0); }, 0);

		var cards = [
			{ label: 'Outstanding', value: money(sum_of(rows)), sub: rows.length + ' accounts', tone: 'total' },
			{ label: 'Unfiled', value: money(sum_of(unassigned)), sub: unassigned.length + ' accounts', tone: 'warn' },
			{ label: 'Filed', value: money(sum_of(assigned)), sub: assigned.length + ' accounts', tone: 'good' },
			{ label: 'Oldest invoice', value: oldest ? oldest + ' days' : '—', sub: 'in view', tone: 'age' }
		];

		$('#arl-cards').html(cards.map(function (c) {
			return `<div class="arl-card arl-card-${c.tone}">
						<span class="arl-card-label">${esc(c.label)}</span>
						<span class="arl-card-value">${esc(c.value)}</span>
						<span class="arl-card-sub">${esc(c.sub)}</span>
					</div>`;
		}).join(''));
	}

	function render_segment_cards() {
		var base = rows_before_segment();
		var active = state.filters.segment;

		var html = all_segments().map(function (seg) {
			var key = seg === '' ? '__unassigned__' : seg;
			var rows = base.filter(function (r) { return (r.segment || '') === seg; });
			var is_active = active === key;
			return `<button class="arl-seg-card arl-sec-${slug(seg)} ${is_active ? 'is-active' : ''}"
						data-segment="${esc(key)}">
						<span class="arl-seg-dot"></span>
						<span class="arl-seg-name">${esc(label_of(seg))}</span>
						<span class="arl-seg-count">${rows.length}</span>
						<span class="arl-seg-amount">${money(sum_of(rows), true)}</span>
					</button>`;
		}).join('');

		$('#arl-seg-cards').html(html);
	}

	function render_active_filters() {
		var f = state.filters;
		var bits = [];
		if (f.search) bits.push(['search', 'Search: "' + f.search + '"']);
		if (f.company) bits.push(['company', 'Company: ' + f.company]);
		if (f.age) bits.push(['age', f.age + '+ days old']);
		if (f.min_amount) bits.push(['min_amount', 'Min ' + money(parseFloat(f.min_amount))]);
		if (f.segment) {
			bits.push(['segment', f.segment === '__unassigned__' ? 'Unassigned only' : f.segment]);
		}
		$('#arl-active').html(bits.length
			? bits.map(function (b) {
				return '<span class="arl-pill" data-clear="' + b[0] + '">' + esc(b[1]) + ' <i>&times;</i></span>';
			}).join('')
			: '');
	}

	// ---- table ---------------------------------------------------------
	function options_html(selected) {
		var out = ['<option value=""' + (selected ? '' : ' selected') + '>— Unassigned —</option>'];
		state.segments.forEach(function (s) {
			out.push('<option value="' + esc(s) + '"' + (s === selected ? ' selected' : '') + '>' +
				esc(s) + '</option>');
		});
		return out.join('');
	}

	function row_html(r) {
		var disabled = state.can_edit ? '' : ' disabled';
		var age_class = (r.age_days || 0) >= 365 ? 'arl-age-old'
			: (r.age_days || 0) >= 180 ? 'arl-age-mid' : '';
		return `
			<tr data-customer="${esc(r.customer)}">
				<td class="arl-cust">
					<a href="/app/customer/${encodeURIComponent(r.customer)}" target="_blank" rel="noopener">${esc(r.customer_name || r.customer)}</a>
					<span class="arl-co">${esc(r.companies || '')}</span>
				</td>
				<td class="arl-seg">
					<select class="arl-select arl-sec-${slug(r.segment)}" data-customer="${esc(r.customer)}"${disabled}>
						${options_html(r.segment || '')}
					</select>
				</td>
				<td class="arl-num">${money(r.outstanding)}</td>
				<td class="arl-inv-col">${r.invoices}</td>
				<td class="arl-oldest ${age_class}">
					${esc(r.oldest || '')}<span class="arl-age">${r.age_days ? r.age_days + 'd' : ''}</span>
				</td>
			</tr>`;
	}

	function section_html(seg) {
		var rows = sorted(visible_rows().filter(function (r) { return (r.segment || '') === seg; }));
		if (state.filters.segment && !rows.length) return '';

		var id = slug(seg);
		var collapsed = !!state.collapsed[seg];
		var body = rows.length
			? `<div class="arl-table-wrap"><table class="arl-table">
					<thead><tr>
						<th>Customer</th><th>Segment</th>
						<th class="arl-num">Outstanding</th><th class="arl-inv-col">Inv</th><th>Oldest</th>
					</tr></thead>
					<tbody>${rows.map(row_html).join('')}</tbody>
				</table></div>`
			: '<div class="arl-empty">Nothing filed here yet.</div>';

		return `
			<section class="arl-section arl-sec-${id} ${collapsed ? 'is-collapsed' : ''}">
				<header class="arl-sec-head" data-toggle="${esc(seg)}">
					<span class="arl-caret">&#9662;</span>
					<h3 class="arl-sec-title">${esc(label_of(seg))}</h3>
					<span class="arl-chip">${rows.length}</span>
					<span class="arl-sec-total">${money(sum_of(rows))}</span>
				</header>
				<div class="arl-sec-body">${body}</div>
			</section>`;
	}

	function render() {
		render_cards();
		render_segment_cards();
		render_active_filters();
		var html = all_segments().map(section_html).join('');
		$('#arl-body').html(html || '<div class="arl-empty arl-empty-page">No accounts match these filters.</div>');
	}

	// ---- data ----------------------------------------------------------
	function load() {
		frappe.call({
			method: 'cannabis_management.cannabis_management.page.ar_legacy.ar_legacy.get_data',
			callback: function (r) {
				if (!r.message) return;
				state.rows = r.message.rows || [];
				state.segments = r.message.segments || [];
				state.companies = r.message.companies || [];
				state.can_edit = !!r.message.can_edit;

				var $co = $('#arl-company').empty().append('<option value="">All companies</option>');
				state.companies.forEach(function (c) {
					$co.append('<option value="' + esc(c) + '">' + esc(c) + '</option>');
				});
				$co.val(state.filters.company);

				render();

				if (r.message.field_missing) {
					page.set_indicator(__('Segment field not installed'), 'orange');
					frappe.msgprint({
						title: __('Segments are read-only here'),
						indicator: 'orange',
						message: __('The AR Legacy segment field has not been installed on this site. Run:<br>' +
							'<code>bench --site &lt;site&gt; execute cannabis_management.cannabis_management.page.ar_legacy.ar_legacy.install_segment_field</code>')
					});
				}
			}
		});
	}

	// ---- events --------------------------------------------------------
	page.main.on('change', '.arl-select', function () {
		var $sel = $(this);
		var customer = $sel.data('customer');
		var segment = $sel.val();
		var row = state.rows.find(function (r) { return r.customer === customer; });
		if (!row) return;

		var previous = row.segment || '';
		row.segment = segment;      // optimistic — the row moves straight away
		render();

		frappe.call({
			method: 'cannabis_management.cannabis_management.page.ar_legacy.ar_legacy.set_segment',
			args: { customer: customer, segment: segment },
			error: function () { row.segment = previous; render(); },
			callback: function () {
				frappe.show_alert({
					message: __('{0} → {1}', [customer, segment || __('Unassigned')]),
					indicator: 'green'
				}, 3);
			}
		});
	});

	page.main.on('click', '.arl-seg-card', function () {
		var key = $(this).data('segment');
		state.filters.segment = state.filters.segment === key ? '' : key;
		render();
	});

	page.main.on('click', '.arl-sec-head', function () {
		var seg = $(this).data('toggle');
		state.collapsed[seg] = !state.collapsed[seg];
		render();
	});

	page.main.on('click', '.arl-pill', function () {
		state.filters[$(this).data('clear')] = '';
		sync_controls();
		render();
	});

	function sync_controls() {
		$('#arl-search').val(state.filters.search);
		$('#arl-company').val(state.filters.company);
		$('#arl-age').val(state.filters.age);
		$('#arl-min').val(state.filters.min_amount);
		$('#arl-sort').val(state.filters.sort);
	}

	$('#arl-age').html(AGE_BUCKETS.map(function (b) {
		return '<option value="' + b.value + '">' + b.label + '</option>';
	}).join(''));
	$('#arl-sort').html(SORTS.map(function (s) {
		return '<option value="' + s.value + '">' + s.label + '</option>';
	}).join(''));

	page.main.find('#arl-search').on('input', function () { state.filters.search = $(this).val(); render(); });
	page.main.find('#arl-company').on('change', function () { state.filters.company = $(this).val(); render(); });
	page.main.find('#arl-age').on('change', function () { state.filters.age = $(this).val(); render(); });
	page.main.find('#arl-min').on('input', function () { state.filters.min_amount = $(this).val(); render(); });
	page.main.find('#arl-sort').on('change', function () { state.filters.sort = $(this).val(); render(); });
	page.main.find('#arl-clear').on('click', function () {
		state.filters = { search: '', segment: '', company: '', age: '', min_amount: '', sort: 'outstanding' };
		sync_controls();
		render();
	});
	page.main.find('#arl-refresh').on('click', function () { load(); });

	load();
};
